tripal_chado/src/Plugin/ChadoBuddy/ChadoPropertyBuddy.php

```diff
@@ -0,0 +1,565 @@
+   *     - projectprop.project_id - (optional) base table primary key value
+   *     - projectprop.type_id - a foreign key to cvterm_id
+   *     - projectprop.value - the value of the property
+   *     - projectprop.rank - optional rank of the property
+   *     - and possibly other columns for some property tables
+   *     - cv.cv_id
+   *     - cv.name
+   *     - cv.definition
+   *     - cvterm.cvterm_id
+   *     - cvterm.cv_id
+   *     - cvterm.name
+   *     - cvterm.definition
+   *     - cvterm.is_obsolete
+   *     - cvterm.is_relationshiptype
+   *     - dbxref.dbxref_id
+   *     - dbxref.db_id
+   *     - dbxref.description
+   *     - dbxref.accession
+   *     - dbxref.version
+   *     - db.db_id
+   *     - db.name
+   *     - db.description
+   *     - db.urlprefix
+   *     - db.url
+   *     - buddy_record = a ChadoBuddyRecord can be used
+   *       in place of or in addition to other keys
+   *
+   * @param array $conditions
+   *   An associative array of the conditions to find the record to update.
+   *   Although the same keys are supported as those indicated for the $values,
+   *   only columns that are part of a unique constraint will be used for the
+   *   database query. e.g. you can't query on the property value.
+   *
+   * @param array $options (Optional)
+   *     - property_table - if the default of $base_table . 'prop' needs to be changed
+   *     - fkey - if the default of $base_table . '_id' needs to be changed
+   *     - pkey - if the default of $property_table . '_id' needs to be changed
+   *
+   * @return bool|ChadoBuddyRecord
+   *   The updated ChadoBuddyRecord will be returned on success, FALSE will be
+   *   returned if no record was found to update.
+   *
+   * @throws Drupal\tripal_chado\ChadoBuddy\Exceptions\ChadoBuddyException
+   *   If an error is encountered.
+   */
+  public function updateProperty(string $base_table, int $record_id, array $values, array $conditions, array $options = []) {
+    $property_table = $options['property_table'] ?? $base_table . 'prop';
+    $fkey = $options['fkey'] ?? $base_table . '_id';
+    $pkey = $options['pkey'] ?? $property_table . '_id';
+
+    $valid_tables = ['cvterm', 'cv', 'dbxref', 'db', $base_table, $property_table];
+    $valid_columns = $this->getTableColumns($valid_tables);
+    $values = $this->dereferenceBuddyRecord($values);
+    $conditions = $this->dereferenceBuddyRecord($conditions);
+    $this->validateInput($values, $valid_columns);
+    $this->validateInput($conditions, $valid_columns);
+
+    $existing_records = $this->getProperty($base_table, $record_id, $conditions, $options);
+    if (count($existing_records) == 0) {
+      return FALSE;
+    }
+    if (count($existing_records) > 1) {
+      throw new ChadoBuddyException("ChadoBuddy updateProperty error, more than one record matched the conditions specified\n".print_r($conditions, TRUE));
+    }
```

Comment:
I think this is okay, because there is a query that happens before the update, and the exception is thrown based on those query results, before anything is changed in the db. Please correct me if that is wrong.